samples/Common.c

```diff
@@ -704,24 +715,70 @@ VOID sampleBandwidthEstimationHandler(UINT64 customData, DOUBLE maximumBitrate)
 VOID sampleSenderBandwidthEstimationHandler(UINT64 customData, UINT32 txBytes, UINT32 rxBytes, UINT32 txPacketsCnt, UINT32 rxPacketsCnt,
                                             UINT64 duration)
 {
-    UNUSED_PARAM(customData);
     UNUSED_PARAM(duration);
-    UNUSED_PARAM(rxBytes);
-    UNUSED_PARAM(txBytes);
+    UINT64 videoBitrate, audioBitrate;
+    BOOL locked = FALSE;
     UINT32 lostPacketsCnt = txPacketsCnt - rxPacketsCnt;
-    UINT32 percentLost = lostPacketsCnt * 100 / txPacketsCnt;
-    UINT32 bitrate = 1024;
-    if (percentLost < 2) {
-        // increase encoder bitrate by 2 percent
-        bitrate *= 1.02f;
-    } else if (percentLost > 5) {
-        // decrease encoder bitrate by packet loss percent
-        bitrate *= (1.0f - percentLost / 100.0f);
-    }
-    // otherwise keep bitrate the same
-
-    DLOGV("received sender bitrate estimation: suggested bitrate %u sent: %u bytes %u packets received: %u bytes %u packets in %lu msec", bitrate,
-          txBytes, txPacketsCnt, rxBytes, rxPacketsCnt, duration / 10000ULL);
+    UINT8 percentLost = (txPacketsCnt > 0) ? (lostPacketsCnt * 100 / txPacketsCnt) : 0;
+    SampleStreamingSession* pSampleStreamingSession = (SampleStreamingSession*) customData;
+    if (pSampleStreamingSession == NULL) {
+        DLOGW("Invalid streaming session (NULL object)");
+        return;
+    }
+
+    // Calculate packet loss
+    DOUBLE currentPacketLoss = (DOUBLE) percentLost;
+    pSampleStreamingSession->twccMetadata.averagePacketLoss =
+        EMA_ACCUMULATOR_GET_NEXT(pSampleStreamingSession->twccMetadata.averagePacketLoss, currentPacketLoss);
+
+    UINT64 currentTimeMs = GETTIME();
+    UINT64 timeDiff = currentTimeMs - pSampleStreamingSession->twccMetadata.lastAdjustmentTimeMs;
+    if (timeDiff < ADJUSTMENT_INTERVAL_SECONDS) {
+        // Too soon for another adjustment
+        return;
+    }
+
+    MUTEX_LOCK(pSampleStreamingSession->twccMetadata.updateLock);
+    locked = TRUE;
+    videoBitrate = pSampleStreamingSession->twccMetadata.currentVideoBitrate;
+    audioBitrate = pSampleStreamingSession->twccMetadata.currentAudioBitrate;
+    if (locked) {
+        MUTEX_UNLOCK(pSampleStreamingSession->twccMetadata.updateLock);
+        locked = FALSE;
+    }
+
+    if (pSampleStreamingSession->twccMetadata.averagePacketLoss <= 5) {
+        // increase encoder bitrate by 5 percent with a cap at MAX_BITRATE
+        videoBitrate = (UINT64) MIN(videoBitrate * 1.05f, MAX_VIDEO_BITRATE_KBPS);
+    } else {
+        // decrease encoder bitrate by average packet loss percent, with a cap at MIN_BITRATE
+        videoBitrate = (UINT64) MAX(videoBitrate * (1.0f - pSampleStreamingSession->twccMetadata.averagePacketLoss / 100.0f), MIN_VIDEO_BITRATE_KBPS);
+    }
+
+    if (pSampleStreamingSession->twccMetadata.averagePacketLoss <= 5) {
```

Comment:
During your testing, did you observed any oscillations between good and bad bitrate? Did it got stable after a packet loss?

Reply:
Yeah. I ran tests with tc to change packet loss rates and then switching it off at times too. I saw a better playback recovery and stability compared to without twcc.